porting/configure_test_fix/Overlay.rst

```diff
@@ -3,82 +3,133 @@
 Overlay file method
 ===================
 
-The UBports rootfs comes with a set of standard configuration files for a number of features such as display scaling, sound, bluetooth and more. These files may not tailored to the needs of your specific device and must therefore to be replaced in order for the feature in question to function as it should. This is done with overlay files, i.e. files that the original files get overwritten with. In other words, you need to rewrite the files in question, making the necessary adjustments for your device, and then see to it that these are incorporated into the build so that they will replace the originals.
+The UBports rootfs provides a set of default configuration files for a number of features such as display scaling, sound, bluetooth and more. These files may not be tailored to the device-specific features of the porting target and may therefore have to be adjusted. Furthermore, it may be also be necessary to add additional configuration files as well as scripts for specific needs. Overlay files provide a mechanism to replace existing or add new files to the file system.
 
-If your build is based on the downloaded GSI and rootfs, installed using the halium-install script, then further adjustment of the build will not be possible. This is because the halium-script does not include any method for injecting overlay files. If, on the other hand, your build is based on Gitlab CI scripts, there is a way to further tweak your port with overlay files. This method is described at the end of this section.
+The actual overlay mechanism for replacing or adding files depends both on the build method and the version of Ubuntu Touch. Using the halium-install script to install a build based on the downloaded GSI and rootfs precludes the use of overlay files since it does not provide a mechanism for that.
 
-.. _Overlay-steps:
+In general, there are two methods of implementing overlay files, one is to directly extract them in the target path potentially overwriting existing files, the other is to place them in a designated path from where they are then mounted onto their destination path during the boot process. The latter method is generally preferable as it allows for delta OTA updates and, starting from Ubuntu Touch 20.04, has become more flexible in terms of overlaying files and directories in vendor and Android images. It will be discussed in detail below before describing how overlay files are included into the build process in full system image builds and builds based on GitLab CI scripts.
```

Comment:
> one is to directly extract them in the target path potentially overwriting existing files,...

I think the documentation is not clear yet about what it means to "directly extract".

It's in the format of the device tarball. Anything in `system/` directory in the tarball [1] will be directly extracted into the on-device rootfs by `system-image-upgrader`.

If this device tarball is built as part of full-system.img port, this portion of the device tarball will contain only `system.img` or `android-rootfs.img` (see [here](https://gitlab.com/ubports/community-ports/jenkins-ci/halium-build-tools/-/blob/main/build-tarball.sh)), and no "directly extracted" overlay file exists.

If, however, it's built as part of "Gitlab CI script-based builds", this directory can also include "directly extracted" overlay files (see [here](https://gitlab.com/ubports/community-ports/android9/volla-phone-x/volla-yggdrasilx/-/blob/master/build.sh#L86-87)). This also means, without modifying the build script, placing files in `overlay/` directory of the git repo will _actually_ make them "directly extracted" overlay, which is currently the case for all GItLab CI script-based ports since mount-based system for them is not available for 16.04. [2]

[1] Not to be confused with Android's `/system`. Oh, I hate this naming...
[2] This is also worries me, as the tarball build scripts stays with each device, which means they're _tens of forks_ of these scripts. I have a feeling that we might have to edit them all...

Reply:
> If this device tarball is built as part of full-system.img port, this portion of the device tarball will contain only system.img or android-rootfs.img (see [here](https://gitlab.com/ubports/community-ports/jenkins-ci/halium-build-tools/-/blob/main/build-tarball.sh)), and no "directly extracted" overlay file exists.

But in this case the `PRODUCT_COPY_FILES` macro will do the equivalent to "extraction" by copying it onto the filesystem image, will it not?

Reply:
Unfortunately, no. We're in Halium/Android build system, and it only produces `system.img`/`android-rootfs.img` (and `halium-boot.img`).

Output paths in `PRODUCT_COPY_FILES` are relative to Android's `$OUT` directory. Files in `$OUT/system/` will be in `/` of `system.img` or in `/system/` of `android-rootfs.img`, both of which results in files being available at `/android/system` after Android image mounting is done.

As I said, don't be confused with device tarball's `/system/` which will be extracted over the root partition on device. As noted earlier, for `system.img` ports the device tarball is created in a step after that.

Reply:
OK, thanks for the detailed explanation (again). I've made corrections accordingly.

What I'm still unclear about is when using the Gitlab CI scripts for xenial: Is there a way to use bind mounts? `/opt/halium-overlay` is only supported starting from your modifications in focal, right? So is the only option to overlay files in xenial by extraction?

Reply:
No, Gitlab CI script ports cannot use bind mounts. On xenial, the overlay mounting code looks only in `/android/system`, and since these devices have the same `android-rootfs.img`, that cannot be done.

So, yeah, on xenial they'll have to stick to the directly extracted method.